File: apk/app/src/main/java/app/momentra/invite/InviteLinkParser.kt
package app.momentra.invite

import android.net.Uri
import java.net.URLDecoder

enum class InviteLinkKind {
    GROUP,
    BUSINESS,
}

data class ParsedInviteLink(
    val kind: InviteLinkKind,
    val token: String,
)

fun parseInviteUri(uri: Uri): ParsedInviteLink? {
    val path = (uri.path ?: "").trimEnd('/')
    val businessSeg = "/business/join/"
    val groupSeg = "/join/"
    return when {
        path.contains(businessSeg) -> {
            val raw = path.substringAfter(businessSeg)
            val token = URLDecoder.decode(raw, Charsets.UTF_8.name()).trim()
            if (token.isBlank()) null else ParsedInviteLink(InviteLinkKind.BUSINESS, token)
        }
        path.contains(groupSeg) -> {
            val raw = path.substringAfter(groupSeg)
            val token = URLDecoder.decode(raw, Charsets.UTF_8.name()).trim()
            if (token.isBlank()) null else ParsedInviteLink(InviteLinkKind.GROUP, token)
        }
        else -> null
    }
}

/** Parses a QR payload: full https URL or a bare hex token (treated as group join). */
fun parseInvitePayload(raw: String): ParsedInviteLink? {
    val t = raw.trim()
    if (t.isEmpty()) return null
    if (t.startsWith("http://", ignoreCase = true) || t.startsWith("https://", ignoreCase = true)) {
        return parseInviteUri(Uri.parse(t))
    }
    if (t.length >= 32 && t.all { c -> c.isLetterOrDigit() }) {
        return ParsedInviteLink(InviteLinkKind.GROUP, t)
    }
    return null
}

File: momentra/momentra/InviteLinkParser.swift
//
//  InviteLinkParser.swift
//  momentra
//

import Foundation

enum InviteLinkKind {
    case group
    case business
}

struct ParsedInviteLink {
    let kind: InviteLinkKind
    let token: String
}

enum InviteLinkParser {
    static func parse(url: URL) -> ParsedInviteLink? {
        let path = url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if path.hasPrefix("business/join/") {
            let raw = String(path.dropFirst("business/join/".count))
            let token = raw.removingPercentEncoding ?? raw
            if token.isEmpty { return nil }
            return ParsedInviteLink(kind: .business, token: token)
        }
        if path.hasPrefix("join/") {
            let raw = String(path.dropFirst("join/".count))
            let token = raw.removingPercentEncoding ?? raw
            if token.isEmpty { return nil }
            return ParsedInviteLink(kind: .group, token: token)
        }
        return nil
    }

    /// QR may contain a full URL or a bare token (treated as group).
    static func parsePayload(_ raw: String) -> ParsedInviteLink? {
        let t = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.isEmpty { return nil }
        if let u = URL(string: t), u.scheme == "http" || u.scheme == "https" {
            return parse(url: u)
        }
        if t.count >= 32, t.allSatisfy({ $0.isLetter || $0.isNumber }) {
            return ParsedInviteLink(kind: .group, token: t)
        }
        return nil
    }
}
